### PR TITLE
refactor: implement delete event type in filepath sync, expand tests

### PR DESCRIPTION
**Pull request by james-milligan**
## This PR

- refactors filepath sync to implement `sync.DELETE`
- prevents EOF error when a file is empty
- updated `Flags.Delete` method to delete all flags from the provided source when the `Flags.Flags` slice is empty
- expanded test cases for filepath sync
- introduced 'control' flags to the `Flags.Delete` tests
- introduced handler for `fsnotify.Chmod` events, in Linux the `fsnotify.Remove` event won't fire until all descriptors are closed, the filepath sync prevents this from happening, so following a `fsnotify.Chmod` event the file is checked for 'deletion' using `os.Stat`

### Related Issues

https://github.com/open-feature/flagd/issues/357

### CODE REVIEW

File: pkg/eval/json_evaluator_model.go [hunk 1]
```diff
@@ -85,9 +85,23 @@ func (f Flags) Update(logger *logger.Logger, source string, ff Flags) map[string
 }
 
 // Delete matching flags from source. The implementation is not thread safe
+// If ff is nil, or, ff.Flags is of length 0, all flags from the given source are deleted
 func (f Flags) Delete(logger *logger.Logger, source string, ff Flags) map[string]interface{} {
 	notifications := map[string]interface{}{}
 
+	if ff.Flags == nil || len(ff.Flags) == 0 {
```
Comment (skyerus) [inline pkg/eval/json_evaluator_model.go:92]:
```suggestion
	if len(ff.Flags) == 0 {
```
if nil, len is also 0

File: pkg/eval/json_evaluator_model.go [hunk 2]
```diff
@@ -85,9 +85,23 @@ func (f Flags) Update(logger *logger.Logger, source string, ff Flags) map[string
 }
 
 // Delete matching flags from source. The implementation is not thread safe
+// If ff is nil, or, ff.Flags is of length 0, all flags from the given source are deleted
```
Comment (skyerus) [inline pkg/eval/json_evaluator_model.go:88]:
```suggestion
// If ff.Flags is empty, all flags from the given source are deleted
```

File: pkg/sync/file/filepath_sync.go [hunk 3]
```diff
@@ -57,22 +54,23 @@ func (fs *Sync) Sync(ctx context.Context, dataSync chan<- sync.DataSync) error {
...
 				// Updates cause a remove event, we need to re-add the watcher in this case.
 				err = watcher.Add(fs.URI)
 				if err != nil {
+					// the watcher could not be re-added, so the file must have been deleted
 					fs.Logger.Error(fmt.Sprintf("error restoring watcher, file may have been deleted: %s", err.Error()))
+					fs.sendDataSync(ctx, sync.DELETE, dataSync)
+					continue
 				}
+
+				// Counterintuively, remove events are the only meanful ones seen in K8s.
```
Comment (skyerus) [inline pkg/sync/file/filepath_sync.go:70]:
```suggestion
				// Counterintuitively, remove events are the only meaningful ones seen in K8s.
```

File: pkg/sync/file/filepath_sync_test.go [hunk 4]
```diff
@@ -6,42 +6,140 @@ import (
...
+						t.Errorf("expected content: %s, but received content: %s", syncEvent.FlagData, data.FlagData)
+					}
+					if data.Source != syncEvent.Source {
+						t.Errorf("expected source: %s, but received source: %s", syncEvent.Source, data.Source)
+					}
+					if data.Type != syncEvent.Type {
+						t.Errorf("expected type: %b, but received type: %b", syncEvent.Type, data.Type)
+					}
+				case <-time.After(10 * time.Second):
+					t.Errorf("event not found, timeout out after 3 seconds")
```
Comment (skyerus) [inline pkg/sync/file/filepath_sync_test.go:139]:
```suggestion
					t.Errorf("event not found, timeout out after 10 seconds")
```

**Review by skyerus — APPROVED** [review]

File: pkg/sync/file/filepath_sync.go [hunk 5]
```diff
@@ -37,14 +40,12 @@ func (fs *Sync) Sync(ctx context.Context, dataSync chan<- sync.DataSync) error {
...
-	// file watcher is ready(and stable), fetch and emit the initial results
-	fetch, err := fs.fetch(ctx)
+	info, err := os.Stat(fs.URI)
 	if err != nil {
-		return err
+		fs.Logger.Error(fmt.Sprintf("error fetching file info: %s %s", fs.URI, err.Error()))
 	}
-
-	dataSync <- sync.DataSync{FlagData: fetch, Source: fs.URI, Type: sync.ALL}
+	fmt.Println(info)
```
Comment (skyerus) [inline pkg/sync/file/filepath_sync.go:47]: remove

File: pkg/sync/file/filepath_sync.go [hunk 6]
```diff
@@ -57,21 +58,29 @@ func (fs *Sync) Sync(ctx context.Context, dataSync chan<- sync.DataSync) error {
...
+					continue
+				}
+
+				// Counterintuitively, remove events are the only meaningful ones seen in K8s.
+				// K8s handles mounted ConfigMap updates by modifying symbolic links, which is an atomic operation.
+				// At the point the remove event is fired, we have our new data, so we can send it down the channel.
+				fs.sendDataSync(ctx, sync.ALL, dataSync)
+			} else if event.Has(fsnotify.Chmod) {
+				// on linux the REMOVE event will not fire until all file descriptors are closed, this cannot happen
+				// whilst we are watching the file, os.Stat is used here to infer deletion
```
Comment (skyerus) [inline pkg/sync/file/filepath_sync.go:80]:
```suggestion
				// while the file is being watched, os.Stat is used here to infer deletion
```

File: pkg/sync/file/filepath_sync.go [hunk 7]
```diff
@@ -37,14 +40,12 @@ func (fs *Sync) Sync(ctx context.Context, dataSync chan<- sync.DataSync) error {
 	if err != nil {
 		return err
 	}
-
-	// file watcher is ready(and stable), fetch and emit the initial results
-	fetch, err := fs.fetch(ctx)
+	info, err := os.Stat(fs.URI)
```
Comment (skyerus) [inline pkg/sync/file/filepath_sync.go:43]: is this used?

File: pkg/eval/json_evaluator_model.go [hunk 8]
```diff
@@ -98,7 +112,7 @@ func (f Flags) Delete(logger *logger.Logger, source string, ff Flags) map[string
 			delete(f.Flags, k)
 		} else {
 			logger.Warn(
-				fmt.Sprintf("failed to remove flag, flag with key %s from source %s does not exisit.",
+				fmt.Sprintf("failed to remove flag, flag with key %s from source %s does not exist.",
```
Comment (Kavindu-Dodan) [inline pkg/eval/json_evaluator_model.go:115]: "leave it cleaner than you found it" 👍

**Review by Kavindu-Dodan — APPROVED** [review]

File: pkg/sync/file/filepath_sync.go [hunk 9]
```diff
@@ -24,6 +24,9 @@ type Sync struct {
 	fileType string
 }
 
+// default state is used to prevent EOF errors when handling filepath delete events + empty files
+const DefaultState = "{}"
```
Comment (skyerus) [inline pkg/sync/file/filepath_sync.go:28]: does this need to be exported?

**Review by toddbaert — APPROVED** [review]

File: pkg/sync/file/filepath_sync.go (hunk 9 above)
Comment (james-milligan) [reply pkg/sync/file/filepath_sync.go]: resolved

**Comment by beeme1mr** [comment]
Hey @james-milligan, could you please fix the merge conflicts?

**Pull request merged**